events/reporter.go

```diff
@@ -0,0 +1,430 @@
+func ReportTxWithValidity(tx *types.Transaction, valid bool) {
+	txWithValidity := TransactionWithValidity{
+		Transaction: tx,
+		Valid:       valid,
+	}
+	if reporter != nil {
+		if reporter.blocking {
+			reporter.channelTransaction <- txWithValidity
+			log.Info("reported tx on channelTransaction: %v", txWithValidity)
+		} else {
+			select {
+			case reporter.channelTransaction <- txWithValidity:
+				log.Info("reported tx on channelTransaction: %v", txWithValidity)
+			default:
+				log.Info("not reporting tx as no one is listening: %v", txWithValidity)
+			}
+		}
+	}
+}
+
+// ReportValidTx reports a valid transaction
+func ReportValidTx(tx *types.Transaction, valid bool) {
+	Publish(ValidTx{ID: tx.ID().String(), Valid: valid})
+}
+
+// ReportNewActivation reports a new activation
+func ReportNewActivation(activation *types.ActivationTx, layersPerEpoch uint16) {
+	Publish(NewAtx{
+		ID:      activation.ShortString(),
+		LayerID: uint64(activation.PubLayerID.GetEpoch(layersPerEpoch)),
+	})
+
+	if reporter != nil {
+		if reporter.blocking {
+			reporter.channelActivation <- activation
+			log.Info("reported activation")
+		} else {
+			select {
+			case reporter.channelActivation <- activation:
+				log.Info("reported activation")
+			default:
+				log.Info("not reporting activation as no one is listening")
+			}
+		}
+	}
+}
+
+// ReportRewardReceived reports a new reward
+func ReportRewardReceived(r Reward) {
+	Publish(RewardReceived{
+		Coinbase: r.Coinbase.String(),
+		Amount:   r.Total,
+	})
+
+	if reporter != nil {
+		if reporter.blocking {
+			reporter.channelReward <- r
+			log.Info("reported reward: %v", r)
+		} else {
+			select {
+			case reporter.channelReward <- r:
+				log.Info("reported reward: %v", r)
+			default:
+				log.Info("not reporting reward as no one is listening: %v", r)
+			}
+		}
+	}
+}
+
+// ReportNewBlock reports a new block
+func ReportNewBlock(blk *types.Block) {
+	Publish(NewBlock{
+		ID:    blk.ID().String(),
+		Atx:   blk.ATXID.ShortString(),
+		Layer: uint64(blk.LayerIndex),
+	})
+}
+
+// ReportValidBlock reports a valid block
+func ReportValidBlock(blockID types.BlockID, valid bool) {
+	Publish(ValidBlock{
+		ID:    blockID.String(),
+		Valid: valid,
+	})
+}
+
+// ReportAtxCreated reports a created activation
+func ReportAtxCreated(created bool, layer uint64) {
+	Publish(AtxCreated{Created: created, Layer: layer})
+}
+
+// ReportValidActivation reports a valid activation
+func ReportValidActivation(activation *types.ActivationTx, valid bool) {
+	Publish(ValidAtx{ID: activation.ShortString(), Valid: valid})
+}
+
+// ReportDoneCreatingBlock reports a created block
+func ReportDoneCreatingBlock(eligible bool, layer uint64, error string) {
+	Publish(DoneCreatingBlock{
+		Eligible: eligible,
+		Layer:    layer,
+		Error:    error,
+	})
+}
+
+// ReportNewLayer reports a new layer
+func ReportNewLayer(layer NewLayer) {
+	if reporter != nil {
```

Comment:
you can just register to the appropriate pubsub channel within the node to receive these events

Reply:
The reporter events and the pubsub events are pretty different. For one thing, pubsub events are all serialized and sent over TCP, whereas the reporter events are native types and kept in-process. I talked more about this here: https://github.com/spacemeshos/SMIPS/issues/21#issuecomment-653199068. Let's discuss if you still have concerns!